### ISSUE TITLE
Problem login with new user

### ISSUE THREAD

**Issue by Cycl0pe** [gophish/gophish#1375]
What version of Gophish are you using?: V 0.7.1 commit 11493024a07d13c2213913b0ef2e37b2aff7e409

Brief description of the issue: I cannot login with a new user created with the admin user.

What are you expecting to see happen? : Be able to connect to the Gophish administration platform with the new user.

What are you seeing happen? When trying to connect to the administration platform, I get the response "The connection was reset" with a status 302.

Please provide any terminal output that may be relevant below:

```
2019/02/25 12:49:36 http: panic serving 127.0.0.1:39262: runtime error: invalid memory address or nil pointer dereference
goroutine 28 [running]:
net/http.(*conn).serve.func1(0xc4201b25a0)
	/usr/local/go/src/net/http/server.go:1726 +0xd0
panic(0xa4c900, 0x10352f0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/jinzhu/gorm.(*Scope).related(0x0, 0x9f9a60, 0xc420ab8ba0, 0xc420a79500, 0x1, 0x1, 0xc420acf4a0)
	/home/gg/go/src/github.com/jinzhu/gorm/scope.go:1058 +0x3a
github.com/jinzhu/gorm.(*Association).Find(0xc420ae4a20, 0x9f9a60, 0xc420ab8ba0, 0xc420ae4a20)
	/home/gg/go/src/github.com/jinzhu/gorm/association.go:19 +0x90
github.com/gophish/gophish/models.(*User).HasPermission(0xc420588f20, 0xaf6130, 0xd, 0xaa7d60, 0xc42026d3b0, 0x0)
	/home/gg/go/src/github.com/gophish/gophish/models/rbac.go:78 +0x1ed
github.com/gophish/gophish/controllers.newTemplateParams(0xc420abb100, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/gg/go/src/github.com/gophish/gophish/controllers/route.go:176 +0x117
github.com/gophish/gophish/controllers.(*AdminServer).Base(0xc420210320, 0xb6b780, 0xc4201658f0, 0xc420abb100)
	/home/gg/go/src/github.com/gophish/gophish/controllers/route.go:223 +0x43
github.com/gophish/gophish/controllers.(*AdminServer).Base-fm(0xb6b780, 0xc4201658f0, 0xc420abb100)
	/home/gg/go/src/github.com/gophish/gophish/controllers/route.go:97 +0x48
net/http.HandlerFunc.ServeHTTP(0xc420552b40, 0xb6b780, 0xc4201658f0, 0xc420abb100)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/gophish/gophish/middleware.RequireLogin.func1(0xb6b780, 0xc4201658f0, 0xc420abb100)
	/home/gg/go/src/github.com/gophish/gophish/middleware/middleware.go:109 +0xaf
net/http.HandlerFunc.ServeHTTP(0xc4202145a0, 0xb6b780, 0xc4201658f0, 0xc420abb100)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc42054e300, 0xb6b780, 0xc4201658f0, 0xc420abaf00)
	/home/gg/go/src/github.com/gorilla/mux/mux.go:212 +0xcd
github.com/gorilla/csrf.(*csrf).ServeHTTP(0xc42018e960, 0xb6b780, 0xc4201658f0, 0xc420abae00)
	/home/gg/go/src/github.com/gorilla/csrf/csrf.go:267 +0x323
net/http.(Handler).ServeHTTP-fm(0xb6b780, 0xc4201658f0, 0xc420abad00)
	/home/gg/go/src/github.com/gophish/gophish/controllers/route.go:143 +0x4d
net/http.HandlerFunc.ServeHTTP(0xc420215a20, 0xb6b780, 0xc4201658f0, 0xc420abad00)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/gophish/gophish/middleware.CSRFExceptions.func1(0xb6b780, 0xc4201658f0, 0xc420abad00)
	/home/gg/go/src/github.com/gophish/gophish/middleware/middleware.go:30 +0xba
net/http.HandlerFunc.ServeHTTP(0xc420215a40, 0xb6b780, 0xc4201658f0, 0xc420abad00)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/gophish/gophish/middleware.GetContext.func1(0xb6b780, 0xc4201658f0, 0xc420abac00)
	/home/gg/go/src/github.com/gophish/gophish/middleware/middleware.go:60 +0x251
net/http.HandlerFunc.ServeHTTP(0xc420215a60, 0xb6b780, 0xc4201658f0, 0xc420abaa00)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x7f6167f72540, 0xc420a71dd0, 0xc420abaa00)
	/home/gg/go/src/github.com/NYTimes/gziphandler/gzip.go:336 +0x1ae
net/http.HandlerFunc.ServeHTTP(0xc42013e030, 0x7f6167f72540, 0xc420a71dd0, 0xc420abaa00)
	/usr/local/go/src/net/http/server.go:1947 +0x44
github.com/gorilla/handlers.loggingHandler.ServeHTTP(0xb668e0, 0xc42054c3c0, 0xb67820, 0xc42013e030, 0xb1b4d0, 0xb6b680, 0xc4208b6460, 0xc420abaa00)
	/home/gg/go/src/github.com/gorilla/handlers/logging.go:45 +0x133
net/http.serverHandler.ServeHTTP(0xc42022cd00, 0xb6b680, 0xc4208b6460, 0xc420abaa00)
	/usr/local/go/src/net/http/server.go:2694 +0xbc
net/http.(*conn).serve(0xc4201b25a0, 0xb6bc40, 0xc42019aa40)
	/usr/local/go/src/net/http/server.go:1830 +0x651
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2795 +0x27b
...
```

Please provide as many steps as you can to reproduce the problem:
 * Step 1 - Create a new user
 * Step 2 - Login with the new user

**Comment by jordan-wright** [comment]
Hi @Cycl0pe,

It looks like this is a bug introduced with the addition of RBAC. Essentially, I'm not setting a role for new users, so it errors out when attempting to fetch that user's role.

The short term fix would be to revert back to a commit before RBAC was introduced. Another option would be to run a manual SQL command to set the role for the new user.

If you're ok with the user being an admin, then in SQLite, this query would be:

```SQL
UPDATE "users" SET "role_id"=(
    SELECT "id" FROM "roles" WHERE "slug"="admin")
WHERE role_id IS NULL;
```

If you want the user to be a "user" account (documentation is coming soon for the differences, but you can see #1333 for more info), just change "admin" to "user".

For MySQL, the query is:

```SQL
UPDATE `users` SET `role_id`=(
    SELECT `id` FROM `roles` AS role_id WHERE `slug`="user")
WHERE role_id IS NULL;
```

This has slightly different quotes.

Right now, I'm working on a full user management API that will replace the current registration handler. I'm tracking this through #1351. This will let you set the role of the user when they are created. I'll keep this issue open until this bug is fixed.

Thanks for bringing this to my attention!

**Comment by binary-blurb** [comment]
Hey man, it looks like this is because when new users are instantiated, they either do not have a role value set or the role value is set to 0 by default when a new user is created. I was able to login with a newly created user by using your SQL query, but modified.

UPDATE `users` SET `role_id`=(SELECT `id` FROM `roles` AS role_id WHERE `slug`="user") WHERE `role_id` IS 0;

I don't believe 0 is the same a NULL in sqlite, but I'm not certain. Some SQL queries and their results below of the process--

sqlite> SELECT * FROM `users` WHERE role_id IS 0;
2|test1|...snip...|...snip...|0
3|test2|...snip...|...snip...|0
sqlite> UPDATE `users` SET `role_id`=(SELECT `id` FROM `roles` AS role_id WHERE `slug`="user") WHERE `role_id` IS 0;
2|test1|...snip...|...snip...|2
3|test2|...snip...|...snip...|2
sqlite> SELECT * FROM `users`;
1|admin|...snip...|...snip...|1
2|test1|...snip...|...snip...|2
3|test2|...snip...|...snip...|2

Notice the admin account doesn't show up by default since it was the 1 role as an admin. You may already knew all of this, but figured it may be helpful to others if NULL == 0. I can login with the newly created accounts just fine now.

**Comment by H4xl0r** [comment]
Still some work goning on here ? 
@jordan-wright could you get me a info how to contact you ?
would like to support the proccess and maybe budget ;.)

**Comment by jordan-wright** [comment]
Hi @H4xl0r,

Thanks for reaching out! You can contact me at hi@getgophish.com. Don't hesitate to reach out, but that said, please know that I prefer to keep all discussions about Gophish issues here on Github to keep things open to the community.

This is still very much in progress. I'm done with the backend work for a full user management API, and am just finishing up the frontend. My time is a bit more limited these days but it's definitely on the way 😄 

When that's done, I'm going to wrap everything up and tag another 0.8 release. 🎊

**Comment by H4xl0r** [comment]
@jordan-wright 
Reason i was trying to contact you:

I need those premissions and differnet user settings !
i would pay you some money to get you some time for the dev. (just tell me what you need).
So that it will be fixed in time. 

There would be more to implement : 

when it comes to users we should have 3 

1 admin (as now role_id 0)
2 user (just view no options / settings role_id 1)

3 user campaign (user bound to an specific camp. just with view access role_id NUMBER OF CAMP generated at start of an campaign , random username random strong pw , dashboard of campagin shows info.(for admins))

The nav template should look for those role_ids and show and hide some menus. 
(user not able to see settings tap etc ) 
(user capaign just able to see his dashbard bound to the campaign)

i will send you an mail and you can also reply there!

**Comment by jordan-wright** [comment]
I'm just about done creating a full user management API that will be accessible for users with the `ModifySystem` permission (basically admin users). Here's what it looks like:

<img width="1438" alt="Screen Shot 2019-05-30 at 9 23 35 PM" src="https://user-images.githubusercontent.com/1317288/58678082-0eb03100-8324-11e9-8d4d-ac273598de86.png">

This solves the problem of separating admin vs. non-admin users which has nice benefits. For example, the very next thing I'm working on once this is done (and the new release is tagged), is integrating Let's Encrypt for domains. Since these domains will be shared across the system, I wanted any admin to be able to manage them (but not just any user!).

The does not solve the problem of view-only access to campaigns. This is because all objects in Gophish are scoped to the current user. So the user could only see campaigns, but they can't make campaigns in the first place so there wouldn't be anything to see 😉 

To solve this problem, I'm going to have to create the concept of "teams". I'm still working through the design, but basically any user could make a team and invite other users. Objects like campaigns, templates, etc. will be shared with the team and users within the team will have certain roles. This is where we could make a read-only role for a team where users could see campaigns, but not make them.

This is going to take a bit to get to, so I'm afraid it's not going to happen soon. I'll certainly keep everyone updated, though!

**Comment by H4xl0r** [comment]
Love it , nice man 👍 👍

**Comment by jordan-wright** [comment]
Closed via 84096b8724b468a5e26329cb009d49a1914474c1.

**Comment by vanditamathur** [comment]
Hi @jordan-wright ,
You have mentioned that you are creating the concept of "teams". Do you have an issue for that which I can look at?
My main requirement is to be able to organize my campaigns for different "teams". Or organize/segregate/group them somehow so that I create/retrieve/update/etc campaigns for a given specific team/group/folder only. Thanks.
